ops/cluster.py

```diff
@@ -21,9 +22,579 @@
+    
+    # Handle cases where gene_symbol_0 might be a column or already the index
+    if 'gene_symbol_0' in df_copy.columns:
+        df_copy = df_copy.set_index('gene_symbol_0')
+    
+    # Fit scaler on control samples
+    scaler = StandardScaler()
+    control_mask = df_copy.index.str.startswith(control_prefix)
+    scaler.fit(df_copy[control_mask].values)
+    
+    # Transform all data
+    df_norm = pd.DataFrame(
+        scaler.transform(df_copy.values),
+        index=df_copy.index,
+        columns=df_copy.columns
+    )
+    
+    return df_norm
+
+def perform_pca_analysis(df, variance_threshold=0.95, save_plot_path=None):
+    """
+    Perform PCA analysis and create explained variance plot.
+    Expects gene_symbol_0 to be the index.
+    
+    Args:
+        df (pd.DataFrame): Data with gene symbols as index
+        variance_threshold (float): Cumulative variance threshold (default 0.95)
+        save_plot_path (str): Path to save variance plot (optional)
+        
+    Returns:
+        tuple: (pca_df, n_components, pca_object)
+            - pca_df: DataFrame with PCA transformed data (gene symbols as index)
+            - n_components: Number of components needed to reach variance threshold
+            - pca_object: Fitted PCA object
+    """
+    # Initialize and fit PCA
+    pca = PCA()
+    pca_transformed = pca.fit_transform(df)
+    
+    # Create DataFrame with PCA results
+    n_components_total = pca_transformed.shape[1]
+    pca_df = pd.DataFrame(
+        pca_transformed,
+        columns=[f'pca_{n}' for n in range(n_components_total)],
+        index=df.index
+    )
+    
+    # Find number of components needed for threshold
+    cumsum = pca.explained_variance_ratio_.cumsum()
+    n_components = np.argwhere(cumsum >= variance_threshold)[0][0] + 1
+    
+    # Create variance plot
+    plt.figure(figsize=(10, 6))
+    plt.plot(cumsum, '-')
+    plt.axhline(variance_threshold, linestyle='--', color='red', 
+                label=f'{variance_threshold*100}% Threshold')
+    plt.axvline(n_components, linestyle='--', color='blue', 
+                label=f'n={n_components}')
+    plt.ylabel('Cumulative fraction of variance explained')
+    plt.xlabel('Number of principal components included')
+    plt.title('PCA Explained Variance Ratio')
+    plt.grid(True)
+    plt.legend()
+    
+    if save_plot_path:
+        plt.savefig(save_plot_path, bbox_inches='tight', dpi=300)
+    plt.close()  # Close the figure to free memory
+    
+    print(f"Number of components needed for {variance_threshold*100}% variance: {n_components}")
+    print(f"Shape of input data: {df.shape}")
+    
+    # Create threshold-limited version
+    pca_df_threshold = pca_df[[f'pca_{i}' for i in range(n_components)]]
+    
+    print(f"Shape of PCA transformed and reduced data: {pca_df_threshold.shape}")
+
+    return pca_df_threshold, n_components, pca
+
+def run_phate(df, random_state=42, n_jobs=4, knn=10, metric='euclidean', **kwargs):
```

Comment:
Oh looks like you did use it here. I think we want to use it wherever possible for sure.